Document the usage of custom optimizers with optimize_prompt (#17084)
mlflow/mlflow: docs/docs/genai/prompt-registry/optimize-prompts.mdx

# Optimize Prompts (Experimental)

MLflow allows you to plug your prompts into advanced prompt optimization techniques through MLflow's unified interface using the <APILink fn="mlflow.genai.optimize_prompt" /> API.

> This feature helps you improve your prompts automatically by leveraging evaluation metrics and labeled data. Built-in support includes [DSPy's MIPROv2 algorithm](https://dspy.ai/api/optimizers/MIPROv2/), and you can also create custom optimization algorithms by extending MLflow's optimizer base class <APILink fn="mlflow.genai.optimize.BasePromptOptimizer" >BasePromptOptimizer</APILink>.

Reviewer: ```suggestion
This feature helps you improve your prompts automatically by leveraging evaluation metrics and labeled data. MLflow provides built-in support for optimization algorithms such as [DSPy’s MIPROv2](https://dspy.ai/api/optimizers/MIPROv2/). You can also implement custom optimization algorithms by extending MLflow's base optimizer class <APILink fn="mlflow.genai.optimize.BasePromptOptimizer" >BasePromptOptimizer</APILink>.
```

> This feature helps you improve your prompts automatically by leveraging evaluation metrics and labeled data. MLflow provides built-in support for optimization algorithms such as [DSPy's MIPROv2 algorithm](https://dspy.ai/api/optimizers/MIPROv2/), and you can also implement custom optimization algorithms by extending MLflow's base optimizer class <APILink fn="mlflow.genai.optimize.BasePromptOptimizer" >BasePromptOptimizer</APILink>.

Reviewer 2: ```suggestion
This feature helps you improve your prompts automatically by leveraging evaluation metrics and labeled data. MLflow provides built-in support for optimization algorithms such as [DSPy's MIPROv2 algorithm](https://dspy.ai/api/optimizers/MIPROv2/). You can also implement custom optimization algorithms by extending MLflow's base optimizer class <APILink fn="mlflow.genai.optimize.BasePromptOptimizer" >BasePromptOptimizer</APILink>.
```

You can customize the optimization process using `OptimizerConfig`, which includes the following parameters:

> - **algorithm**: The optimization algorithm to use. Can be a string (e.g., "DSPy/MIPROv2") or a custom optimizer class. Default: "DSPy/MIPROv2"

Author: Hallucination, it should be class itself

- **num_instruction_candidates**: The number of candidate instructions to try. Default: 6
- **max_few_show_examples**: The maximum number of examples to show in few-shot demonstrations. Default: 6
- **optimizer_llm**: The LLM to use for optimization. Default: None (uses target LLM)
- **extract_instructions**: Whether to extract instructions from the initial prompt template. Default: True
- **verbose**: Whether to show optimizer logs during optimization. Default: False
- **autolog**: Whether to log the optimization parameters, datasets and metrics. If set to True, a MLflow run is automatically created to store them. Default: False

See <APILink fn="mlflow.genai.OptimizerConfig" /> for more details.

## Custom Optimizers

MLflow supports creating custom prompt optimization algorithms by extending the base optimizer classes. This allows you to implement domain-specific optimization strategies or integrate with other optimization libraries.

### BasePromptOptimizer

> For completely custom optimization logic, extend the <APILink fn="mlflow.genai.optimize.BasePromptOptimizer" >BasePromptOptimizer</APILink> class:

Reviewer: Do we need `completely`?

Author: I thought it's clear to say "completely" to compare BasePromptOptimizer and DSPyPromptOptimizer

```python
from mlflow.genai.optimize import BasePromptOptimizer, OptimizerConfig, OptimizerOutput
from mlflow.genai.optimize.types import LLMParams
from mlflow.entities.model_registry import PromptVersion
from typing import Optional
import pandas as pd

class CustomOptimizer(BasePromptOptimizer):
    def optimize(
        self,
        prompt: PromptVersion,
        target_llm_params: LLMParams,
        train_data: pd.DataFrame,
        scorers: list,

>         objective: Optional[callable] = None,

Reviewer 3: ```suggestion
        objective: Optional[Callable[...]] = None,
```

        eval_data: Optional[pd.DataFrame] = None,
    ) -> OptimizerOutput:
        # Implement your custom optimization logic here
        # This is where you would run your optimization algorithm

        optimized_template = f"Please answer accurately: {prompt.template}"

        return OptimizerOutput(
            optimized_prompt=optimized_template,
            optimizer_name="CustomOptimizer",
            final_eval_score=0.85,
            initial_eval_score=0.75,
        )

# Use the custom optimizer
result = mlflow.genai.optimize_prompt(
    target_llm_params=LLMParams(model_name="openai/gpt-4o-mini"),
    prompt=prompt,
    train_data=train_data,
    scorers=[exact_match],
    optimizer_config=OptimizerConfig(algorithm=CustomOptimizer),
)
```

### DSPyPromptOptimizer

For DSPy-based optimizations, extend the <APILink fn="mlflow.genai.optimize.DSPyPromptOptimizer" >DSPyPromptOptimizer</APILink> class, which provides DSPy integration infrastructure:

```python
from mlflow.genai.optimize import (
    DSPyPromptOptimizer,
    OptimizerOutput,
    format_dspy_prompt,
)
import dspy

class CustomDSPyOptimizer(DSPyPromptOptimizer):

>     def run_optimization(

Reviewer: Can we make it clear that users need to implement this method?

    def run_optimization(
        self,
        prompt: PromptVersion,
        program: dspy.Module,
        metric: Callable[[dspy.Example], float],
        train_data: list,
        eval_data: list,
    ) -> OptimizerOutput:
        # Use DSPy's optimization components with your custom logic
        # Example using DSPy's BootstrapFewShot optimizer
        optimizer = dspy.BootstrapFewShot(
            metric=metric,
            max_bootstrapped_demos=self.optimizer_config.max_few_show_examples,
        )

        # Compile the program
        compiled_program = optimizer.compile(
            student=program,
            trainset=train_data,
        )

        return OptimizerOutput(
            optimized_prompt=format_dspy_prompt(compiled_program),
            optimizer_name="BootstrapFewShot",
        )

# Use the custom DSPy optimizer
result = mlflow.genai.optimize_prompt(
    target_llm_params=LLMParams(model_name="openai/gpt-4o-mini"),
    prompt=prompt,
    train_data=train_data,
    scorers=[exact_match],
    optimizer_config=OptimizerConfig(algorithm=CustomDSPyOptimizer),
)
```

:::note

> When using custom optimizers, ensure they return an `OptimizerOutput` object with the optimized prompt and evaluation scores. The DSPy-based optimizer provides additional utilities for working with DSPy's ecosystem of optimization techniques.

Reviewer 2: The 2nd sentence here seems unrelated to the note callout. It might be better to replace that sentence with one that explains why it's critical to return the OptimizerOutput that contains those attributes.

    objective=objective,
)
```

### How to create custom optimizers?

MLflow provides two base classes for creating custom optimizers:

> 1. **<APILink fn="mlflow.genai.optimize.BasePromptOptimizer" >BasePromptOptimizer</APILink>**: For custom optimization logic. You implement the `optimize` method.

Reviewer 2: ```suggestion
1. **<APILink fn="mlflow.genai.optimize.BasePromptOptimizer" >BasePromptOptimizer</APILink>**: For custom optimization logic. In order to customize, you must implement your own `optimize` method.
```

> 2. **<APILink fn="mlflow.genai.optimize.DSPyPromptOptimizer" >DSPyPromptOptimizer</APILink>**: For DSPy-based optimizations. You implement the `run_optimization` method and MLflow handles DSPy setup.

Reviewer 2: ```suggestion
2. **<APILink fn="mlflow.genai.optimize.DSPyPromptOptimizer" >DSPyPromptOptimizer</APILink>**: For DSPy-based optimizations. Customization of this optimizer involves building a`run_optimization` method. MLflow will handle the required DSPy setup for utilizing your custom interface.
```
